src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,91 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        //
+        // where op = `*` `/` or `+`.
+        if ((matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, any(&x), unary(Neg, any(&y))),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, any(&x), unary(Neg, any(&y))),
+                            fval(&c)))) &&
+            (op == Abstract::getBinary(bin->type, Mul) ||
+             op == Abstract::getBinary(bin->type, DivS) ||
+             op == Abstract::getBinary(bin->type, Add))) {
+          c->value = c->value.neg();
+          bin->left = x;
+          bin->right = y;
+          return curr;
+        }
+        // x + fneg(y)   ==>   x - y
+        if (matches(curr, binary(&bin, Add, any(), unary(Neg, any(&y)))) &&
+            !y->is<Const>()) {
```

Comment:
why disallow a constant here? same questions as before - if this is important, we need a comment explaining why. likewise in the cases below that disallow a constant.

Reply:
Because if y is `NaN` const. It will change sign and return different result. Example: https://webassembly.studio/?f=pw5v4w0zpz

Probably make sense hide this under `fastMath`. Wdyt?

Reply:
But it's never true to say "this rule works on everything but a constant", because the value at runtime will be a constant (which could be a NaN constant). In other words: if it's false for a constant, it's false in general. Or am I misreading your statement?

Sorry, I'm not sure what to do with that link. I can build and run, but how do I see the wat before and after this optimization (assuming that's what you want me to see)? Anyhow, would be simpler without a link, just a simple testcase here.

Reply:
> In other words: if it's false for a constant, it's false in general. Or am I misreading your statement?

It's true. Except for NaN in wasm spec if I'm not mistaken. In wasm spec handling with constant NaNs should be always deterministic. But shouldn't guarantee this for general case and arithmetic operations when result with NaN usually always non-deterministic.

Reply:
here test case:
```wat
  (func $uopt (export "uopt") (type $t0) (param $p0 f32) (param $p1 f32) (result i32)
    get_local $p0
    get_local $p1
    f32.neg
    f32.add
    i32.reinterpret/f32
  )

  (func $opt (export "opt") (type $t0) (param $p0 f32) (param $p1 f32) (result i32)
    get_local $p0
    get_local $p1
    f32.sub
    i32.reinterpret/f32
  )
```

here results for `uopt(1.0, NaN)` and `opt(1.0, NaN)`:

```
uopt(1.0, NaN): 0xffc00000
opt(1.0, NaN):  0x7fc00000
```
So, sign of NaNs different before and after optimization which not valid according to spec

Reply:
What do you mean by this:

> In wasm spec handling with constant NaNs should be always deterministic.

?

Maybe I'm misunderstanding you, but I don't think it's true that the wasm spec says math on NaN constants must be deterministic. E.g. `1 + NaN` - that is, an f32 add on two constants, one of whom is a NaN - is nondeterministic, and can result in any arithmetic NaN (just the same as if the NaN were loaded from a local and not a constant in the binary).

Reply:
I mean this spec interpreter's comment:

> * When the result of a binary operation is NaN, the resulting NaN is computed
> * from one of the NaN inputs, if there is one. If both are NaN, one is
> * selected nondeterminstically. If neither, we use a default NaN value.

[source](https://github.com/WebAssembly/spec/blob/master/interpreter/exec/float.ml#L94)

So when one of arguments is NaN it deterministically picked this nan, but if Both NaN it's not deterministic any more and we shouldn't care about type of NaN in this case I guess. Or I misunderstanding this part?

Reply:
Ok. I removed check for const and hide under `fastMath` flags

Reply:
@MaxGraey I think even in that interpreter code, there is no special handling of whether something is a an `*.const` in wasm. Everything the interpreter sees is a constant value. So there is no special-casing of constants known in the wasm vs at runtime.

Otherwise I think you understood that interpreter code correctly. But the interpreter is implementing one possible option out of many others that are all valid under wasm. So the interpreters behavior is not the same as the wasm spec. See https://www.w3.org/TR/wasm-core-1/#nan-propagation%E2%91%A0

Reply:
I see. So spec guarantee deterministic behaviour only for  fneg, fabs, or fcopysign.

Alright. Could you check latest changes. I hided all rules which may lead to change NaN sign under the `fastMath`